src/types/chat.rs

```diff
@@ -0,0 +1,445 @@
+use std::collections::{BTreeMap, BTreeSet};
+use std::error::{Error, FromError};
+use std::io;
+
+use rustc_serialize::json;
+use rustc_serialize::json::{Json, ToJson};
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum ChatJsonError {
+    MalformedJson(json::ParserError),
+    IoError(io::Error),
+    NotAnObject,
+    InvalidFieldType,
+    UnknownField(String),
+    InvalidColor(String),
+    InvalidClickEvent,
+    InvalidHoverEvent
+}
+
+impl FromError<io::Error> for ChatJsonError {
+    fn from_error(err: io::Error) -> ChatJsonError {
+        ChatJsonError::IoError(err)
+    }
+}
+
+impl FromError<json::ParserError> for ChatJsonError {
+    fn from_error(err: json::ParserError) -> ChatJsonError {
+        if let json::ParserError::IoError(e) = err {
+            ChatJsonError::IoError(e)
+        } else {
+            ChatJsonError::MalformedJson(err)
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct ChatJson {
+    pub msg: Message,
+    pub extra: Option<Vec<Json>>,
```

Comment:
IIRC we have to use the old format in the UpdateScore packet... *..searching..*

Yes, confirmed, there's http://wiki.vg/Protocol#Chat_Message which states:

> Identifying the difference between Chat/System Message is important as it helps respect the user's chat visibility options. While Position 2 accepts json formatting it will not display, old style formatting works

Position 2 = above action bar

---

**Aside:** There's also packet [0x3E Teams](http://wiki.vg/Protocol#Teams) which uses same colors for Teams as for Chat. Maybe we should move color into `types::consts`.